llvm/lib/Target/AIE/AIECombinerHelper.cpp

```diff
@@ -1109,3 +1111,210 @@ void llvm::applyPadVector(MachineInstr &MI, MachineRegisterInfo &MRI,
   B.buildInstr(AIE2::G_AIE_PAD_VECTOR_UNDEF, {DstReg}, {MatchedInputVector});
   MI.eraseFromParent();
 }
+
+/// Match something like this:
+///  %68:_(s32) = G_CONSTANT i32 0
+///  %93:_(s32) = G_CONSTANT i32 1
+///  %209:_(<16 x s64>) = G_INTRINSIC intrinsic(@llvm.aie2.concat.1024.512.acc),
+///                   %206(<8 x s64>), %208(<8 x s64>)
+///  %216:_(<8 x s64>) = G_INTRINSIC intrinsic(@llvm.aie2.ext.512.1024.acc),
+///                   %209(<16 x s64>), %68(s32)
+///  %219:_(<8 x s64>) = G_INTRINSIC intrinsic(@llvm.aie2.ext.512.1024.acc),
+///                    %209(<16 x s64>), %93(s32)
+
+/// To convert to:
+///   %216:_(<8 x s64>) = COPY %206(<8 x s64>)
+///   %219:_(<8 x s64>) = COPY %208(<8 x s64>)
+bool llvm::matchExtractConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              const AIEBaseInstrInfo &TII,
+                              Register &MatchInfo) {
+
+  GIntrinsic &GMI = cast<GIntrinsic>(MI);
+  std::optional<const AIEBaseInstrInfo::VExtractOpInfo> ExtractOp =
+      TII.getVExtractOpInfo(GMI.getIntrinsicID());
+
+  if (!ExtractOp)
+    return false;
+
+  auto Cst = getIConstantVRegValWithLookThrough(
+      GMI.getOperand(ExtractOp->SubVectorIndex).getReg(), MRI);
+
+  if (!Cst)
+    return false;
+
+  const unsigned ExtractSize =
+      MRI.getType(MI.getOperand(0).getReg()).getSizeInBits();
+
+  MachineInstr &SrcMI =
+      *MRI.getVRegDef(GMI.getOperand(ExtractOp->Src).getReg());
+
+  Register SrcReg;
+  unsigned ConcatSize = 0;
+  const unsigned Index = Cst->Value.getZExtValue();
+  if (SrcMI.getOpcode() == TargetOpcode::G_CONCAT_VECTORS) {
+    SrcReg = SrcMI.getOperand(Index + 1).getReg();
+    ConcatSize = MRI.getType(SrcReg).getSizeInBits();
+
+  } else if (SrcMI.getOpcode() == TargetOpcode::G_INTRINSIC) {
+    GIntrinsic &SrcMIIntrinsic = cast<GIntrinsic>(SrcMI);
+
+    if (const auto ConcatOp =
+            TII.getVConcatOpInfo(SrcMIIntrinsic.getIntrinsicID())) {
```

Comment:
How about supporting `G_CONCAT_VECTOR` in `TII.getVConcatOpInfo`? Then there's a single test to do here

Reply:
I was thinking about, and this raises 2 design decisions:
1 -  If we handle the  `G_CONCAT_VECTOR` (common to all subtargets)  in the base class, we will loose the ability to detect that the implementation is missing for new subtargets (we will remove the `unreachable`).
2 - If we implement this in the target derived class, we should replicate the implementation in all targets, or at least we can have a method that only detects `G_CONCAT_VECTOR` in the base class, that can be called by `getVConcatOpInfo` in the subtarget implementation.

What do you think?

Reply:
I think that I'd have a base version of `getVConcatOpInfo` that only understands `G_CONCAT_VECTOR`, and hope that future versions don't even introduce all those upd/ext/concat intrinsics ;)
FYI @konstantinschwarz @abnikant